src/BuildScriptGenerator/Node/NodePlatform.cs

```diff
@@ -181,9 +182,18 @@ public PlatformDetectorResult Detect(RepositoryContext context)
             if (ctx.SourceRepo.FileExists(NodeConstants.YarnLockFileName))
             {
                 packageManagerCmd = NodeConstants.YarnCommand;
-                packageInstallCommand = NodeConstants.YarnPackageInstallCommand;
                 configureYarnCache = true;
                 packageInstallerVersionCommand = NodeConstants.YarnVersionCommand;
+                // yarn.lock file is valid YAML in Yarn 2+ and .yarnrc.yml file replaces .yarnrc in Yarn 2+.
+                // Applying yarn 2 cache folder name and package install command.
+                if (nodePlatformDetectorResult.HasYarnrcYmlFile
+                    || nodePlatformDetectorResult.IsYarnLockFileValidYamlFormat) {
```

Comment:
An alternative to this check is to test if the content of `yarn.lock` matches this regex
```regexp
/^__metadata:$/m
```